lib/spack/spack/package.py

```diff
@@ -1420,6 +1524,43 @@ def rpath_args(self):
         return " ".join("-Wl,-rpath,%s" % p for p in self.rpath)
 
 
+class Package(PackageBase):
+    phases = ['install', 'log']
+    # This will be used as a registration decorator in user
+    # packages, if need be
+    PackageBase.sanity_check('install')(PackageBase.sanity_check_prefix)
+
+
+class AutotoolsPackage(PackageBase):
+    phases = ['autoreconf', 'configure', 'build', 'install', 'log']
+
+    def autoreconf(self, spec, prefix):
+        """Not needed usually, configure should be already there"""
+        pass
+
+    @PackageBase.sanity_check('autoreconf')
+    def is_configure_or_die(self):
+        if not os.path.exists('configure'):
+            raise RuntimeError('configure script not found in {0}'.format(os.getcwd()))
+
+    def configure_args(self):
+        return list()
```

Comment:
@citibeth `preconditions` and `sanity_checks` are hooks to register functions that will be executed before and after a phase. Their main point at this stage is to showcase that it is possible to have that fine grain of control exposed to who needs it with a nice (at least I hope) syntax.

That said, if a package has `sanity_checks` or `preconditions` that fails then this mean that ether : 
- the check is wrong
- something didn't work correctly and was detected

In the first case the check needs to be fixed, in the second the failure needs to be a failure. Technically it won't be a problem to add a way of preventing checks, but to me it would feel like turning off compiler warnings because there were warnings...

>  Spack does not and I never encountered a situation where they would have helped. 

I don't agree, because whether you call this `sanity_check` or not the following in `package.py`: 
```python
# Ensure that something was actually installed.
if 'install' in self.install_phases:
    self.sanity_check_prefix()
```
is a sanity check, and I don't see valid reasons why you want to skip it.